### PR TITLE
SIR based covid19 prediction model: add initial scripts

### PR DESCRIPTION
**Pull request by wangfan860**
### New Features

- Develop discrete SIR model
- Estimating model parameters using Bayesian inference with MCMC
- 45 days forecasting with choice of 3 future scenarios

### CODE REVIEW

**Review by kmhernan — APPROVED** [review]

File: covid19-notebooks/covid19-notebook-etl-Dockerfile [hunk 1]
```diff
@@ -11,13 +11,18 @@ RUN apt-get update && apt-get install -y \
...
-	python3-scipy
+	python3-scipy \
+	g++
 
 RUN pip3 install awscli==1.18.*
 COPY ./covid19-notebook-etl-requirements.txt /src/
 RUN pip3 install -r covid19-notebook-etl-requirements.txt
+RUN conda update --yes conda && \
+    conda install --quiet --yes ipython=7.29.0 scikit-learn=1.0.1 jupyterlab=3.2.3 aesara=2.3.5 pygpu=0.7.6  && \
+    conda install --yes -c conda-forge arviz=0.5.1 pymc3=3.8 mkl-service=2.4.0 sunode=0.2.1
```
Comment (paulineribeyre) [inline covid19-notebooks/covid19-notebook-etl-Dockerfile:22]: can these be added to `covid19-notebook-etl-requirements.txt` instead?

File: covid19-notebooks/covid19-notebook-etl-run.sh [hunk 2]
```diff
@@ -22,6 +21,9 @@ for file in "${NOTEBOOKS[@]}"; do
   jupyter nbconvert --to notebook --inplace --execute "$file"
 done
 
+echo "Running SIR-based Bayesian model script..."
+python3 covid19-sir-bayes-model/sir_model.py
```
Comment (paulineribeyre) [inline covid19-notebooks/covid19-notebook-etl-run.sh:25]: looking at the Dockerfile, i think you might need `sir_model.py` here instead of `covid19-sir-bayes-model/sir_model.py` 🤔 does this work?

Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-run.sh:25]: or, right, will change it 👍

File: covid19-notebooks/covid19-notebook-etl-Dockerfile (hunk 1 above)
Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-Dockerfile:22]: For the second line, maybe it could work, let me try locally. However, the last line couldn't be combined to `covid19-notebook-etl-requirements.txt`, they are available only in conda's -conda-forge channel, I tried pip and it never worked.

Comment (kmhernan) [reply covid19-notebooks/covid19-notebook-etl-Dockerfile:22]: isn't it more approiate to create an `environment.yml` for conda to build the environment?

Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-Dockerfile:22]: yeah, I'll do that and have some tests ! @kmhernan

Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-Dockerfile]: Now it is activating a conda environment yaml in Dockerfile

File: covid19-notebooks/covid19-notebook-etl-requirements.txt [hunk 3]
```diff
@@ -4,3 +4,7 @@
 google-cloud-bigquery[bqstorage,pandas]>=2.13.1,<3.0.0
 jupyter>=1.0.0,<2.0.0
 seaborn>=0.11.1,<1.0.0
+matplotlib==3.4.3
+plotly==5.5.0
+scipy==1.7.2
+awscli==1.18.*
```
Comment (paulineribeyre) [inline covid19-notebooks/covid19-notebook-etl-requirements.txt:10]: we have 3 ways to install requirements now (directly in the dockerfile, in this requirements file and in the conda config file), and some libraries are installed more than one way, that doesn't sound like a good idea 😬 if [this comment](https://github.com/uc-cdis/covid19-tools/blob/2bb91c26e9caea68271907ebc4dd4045be706096/covid19-notebooks/covid19-notebook-etl-requirements.txt#L1-L2) is not accurate anymore we should move them out of the dockerfile. And if we could have all the requirements in one place that would be better. Not sure how much work that is...

File: covid19-notebooks/requirements.txt [hunk 4]
```diff
@@ -1,10 +1,10 @@
 fastai==1.0.61
 gen3
-matplotlib==3.4.3
-numpy==1.21.4
-pandas==1.3.4
-plotly==5.5.0
+matplotlib
```
Comment (paulineribeyre) [inline covid19-notebooks/requirements.txt:3]: why did you remove the pinned versions?
it's fine but we'll need to test the new [jupyter-covid19 image](https://github.com/uc-cdis/containers/blob/master/jupyter-covid19/Dockerfile) and see if the notebooks still work

File: covid19-notebooks/covid19-notebook-etl-requirements.txt (hunk 3 above)
Comment (paulineribeyre) [inline covid19-notebooks/covid19-notebook-etl-requirements.txt:10]: the snyk check is failing because of vulns in this version of awscli. I think we would be fine updating it, all we use it for is an `aws s3 cp` command. The latest version is 1.22.51, we can try `awscli>=1.22.0,<2.0.0` instead of 1.18

**Review by paulineribeyre — COMMENTED** [review]
need to fix some formatting https://github.com/uc-cdis/covid19-tools/pull/251#issuecomment-1022840913

File: covid19-notebooks/covid19-notebook-etl-Dockerfile [hunk 5]
```diff
@@ -11,13 +11,40 @@ RUN apt-get update && apt-get install -y \
...
+
+# Make conda activate command available from /bin/bash --login shells
+RUN echo ". $CONDA_DIR/etc/profile.d/conda.sh" >> ~/.profile
+
+# Make conda activate command available from /bin/bash --interative shells
+RUN conda init bash
+
+# Deploy conda environment
+COPY environment.yml .
+RUN conda env create -f environment.yml
```
[the remark this answers is not in the document]
Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-Dockerfile:40]: I renamed this file as recommended 😊

File: covid19-notebooks/requirements.txt (hunk 4 above)
Comment (wangfan860) [reply covid19-notebooks/requirements.txt:3]: I didn't touch this file. If you compare the [master branch](https://github.com/uc-cdis/covid19-tools/blob/master/covid19-notebooks/requirements.txt) and [my branch](https://github.com/uc-cdis/covid19-tools/blob/feat/sir-bayes/covid19-notebooks/requirements.txt), they are the same. Don't know why GitHub shows the change.

File: covid19-notebooks/covid19-notebook-etl-requirements.txt (hunk 3 above)
Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-requirements.txt:10]: I removed the `covid19-notebook-etl-requirements.txt` file and put all the requirements in `covid19-notebook-etl-environment.yml` to have everything in one place. The image was built successfully without any problem. The `python3 sir_model.py` can run as well.😉

Comment (wangfan860) [reply covid19-notebooks/covid19-notebook-etl-requirements.txt:10]: I changed the version of `awscli` in `covid19-notebook-etl-environment.yml` file, there is no problem during docker building. Everything runs fine based on the new image.

File: covid19-notebooks/requirements.txt (hunk 4 above)
Comment (paulineribeyre) [reply covid19-notebooks/requirements.txt]: oh then could you rebase your branch on master

Comment (wangfan860) [reply covid19-notebooks/requirements.txt]: I rebased my branch on top of master, and this change disappeared

File: covid19-notebooks/covid19-sir-bayes-model/sir_model.py [hunk 6]
```diff
@@ -503,7 +503,7 @@ def return_obs_cases_future(trace):
     for label, color, legend in zip(obs_cases_labels_local, colors, legends_list[1]):
         time = np.arange(0, num_days_to_predict)
         cases = dict_obsc_cases[label]
-        cases = cases + cases_obs[0]
+        cases = cases + cases_obs[0] * 0.95
```
[the remark this answers is not in the document]
Comment (wangfan860) [reply covid19-notebooks/covid19-sir-bayes-model/sir_model.py:506]: I'll let Kyle review this again once we figure out the docker build stuff

**Review by paulineribeyre — APPROVED** [review]

**Review by kmhernan — APPROVED** [review]
seems fine to me
